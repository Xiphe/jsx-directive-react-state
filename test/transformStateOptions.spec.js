const transformStateOptions = require('../src/transformStateOptions');
const babel = require('babel-core');
const errorMatchingSnapshot = require('./helpers/errorMatchingSnapshot');

function transform(input) {
  const sourceString = `<div options="${input}" />`;

  const code = babel.transform(
    sourceString,
    {
      filename: 'test.js',
      babelrc: false,
      plugins: [
        'syntax-jsx',
        () => ({
          visitor: {
            JSXAttribute(path) {
              const value = path.get('value');

              try {
                value.replaceWith(transformStateOptions(babel, value.node));
              } catch (e) {
                throw value.hub.file.buildCodeFrameError(value.node, e.message);
              }
            },
          },
        }),
      ],
    },
  ).code;

  return code;
}

describe('transformStateOptions', () => {
  it('transforms a simple state getter', () => {
    expect(transform('foo')).toMatchSnapshot();
  });

  it('transforms a state getter with name mapping', () => {
    expect(transform('foo:bar')).toMatchSnapshot();
  });

  it('throws a syntax error when multiple name mappings are used', () => {
    expect(() => {
      transform('foo:bar:baz');
    }).toThrow(errorMatchingSnapshot());
  });

  it('transforms a state getter with scope', () => {
    expect(transform('foo/bar')).toMatchSnapshot();
  });

  it('throws a syntax error when scope is deeper than 1', () => {
    expect(() => {
      transform('foo/bar/baz');
    }).toThrow(errorMatchingSnapshot());
  });

  it('transforms a state getter with scope and name mapping', () => {
    expect(transform('aww:foo/bar')).toMatchSnapshot();
  });

  it('throws a correct code frame when scope is deeper than 1 with name mapping', () => {
    expect(() => {
      transform('drrt:foo/bar/baz');
    }).toThrow(errorMatchingSnapshot());
  });

  it('transforms a state setter', () => {
    expect(transform('toggle(foo)')).toMatchSnapshot();
  });

  it('throws a syntax error when setters are nested', () => {
    expect(() => {
      transform('foo(bar())');
    }).toThrow(errorMatchingSnapshot());
  });

  it('throws a syntax error setter is not closed', () => {
    expect(() => {
      transform('foo(bar');
    }).toThrow(errorMatchingSnapshot());
  });

  it('throws a syntax error setter is not end of statement', () => {
    expect(() => {
      transform('foo(bar)lol');
    }).toThrow(errorMatchingSnapshot());
  });

  it('transforms a state setter with name mapping and scope', () => {
    expect(transform('onClick:toggle(foo/bar)')).toMatchSnapshot();
  });

  it('throws a correct code frame when scope is deeper than 1 with name mapping setter', () => {
    expect(() => {
      transform('drrt:brr(foo/bar/baz)');
    }).toThrow(errorMatchingSnapshot());
  });

  it('transforms two state declarations', () => {
    expect(transform('value:name;onChange:set(name)')).toMatchSnapshot();
  });

  it('throws a correct code frame for second state getter', () => {
    expect(() => {
      transform('value:name;other:value:name');
    }).toThrow(errorMatchingSnapshot());
  });

  it('throws error when keys are used twice', () => {
    expect(() => {
      transform('foo;foo');
    }).toThrow(errorMatchingSnapshot());
  });

  it('throws error when keys are used twice via name mapping', () => {
    expect(() => {
      transform('set(name);set:foo');
    }).toThrow(errorMatchingSnapshot());
  });
});
