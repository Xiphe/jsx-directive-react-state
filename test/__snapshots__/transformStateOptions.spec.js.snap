// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transformStateOptions throws a correct code frame for second state getter 1`] = `
"test.js: Unexpected token :
> 1 | <div options=\\"value:name;other:value:name\\" />
    |                                     ^"
`;

exports[`transformStateOptions throws a correct code frame when scope is deeper than 1 with name mapping 1`] = `
"test.js: Unexpected token /
> 1 | <div options=\\"drrt:foo/bar/baz\\" />
    |                           ^"
`;

exports[`transformStateOptions throws a correct code frame when scope is deeper than 1 with name mapping setter 1`] = `
"test.js: Unexpected token /
> 1 | <div options=\\"drrt:brr(foo/bar/baz)\\" />
    |                               ^"
`;

exports[`transformStateOptions throws a syntax error setter is not closed 1`] = `
"test.js: Unexpected end of input
> 1 | <div options=\\"foo(bar\\" />
    |                      ^"
`;

exports[`transformStateOptions throws a syntax error setter is not end of statement 1`] = `
"test.js: Expected end of input
> 1 | <div options=\\"foo(bar)lol\\" />
    |                       ^"
`;

exports[`transformStateOptions throws a syntax error when multiple name mappings are used 1`] = `
"test.js: Unexpected token :
> 1 | <div options=\\"foo:bar:baz\\" />
    |                      ^"
`;

exports[`transformStateOptions throws a syntax error when scope is deeper than 1 1`] = `
"test.js: Unexpected token /
> 1 | <div options=\\"foo/bar/baz\\" />
    |                      ^"
`;

exports[`transformStateOptions throws a syntax error when setters are nested 1`] = `
"test.js: Unexpected token (
> 1 | <div options=\\"foo(bar())\\" />
    |                      ^"
`;

exports[`transformStateOptions throws error when keys are used twice 1`] = `
"test.js: Key 'foo' has already been declared
> 1 | <div options=\\"foo;foo\\" />
    |                   ^"
`;

exports[`transformStateOptions throws error when keys are used twice via name mapping 1`] = `
"test.js: Key 'set' has already been declared
> 1 | <div options=\\"set(name);set:foo\\" />
    |                         ^"
`;

exports[`transformStateOptions transforms a simple state getter 1`] = `
"<div options={[{
  key: \\"foo\\"
}]} />;"
`;

exports[`transformStateOptions transforms a state getter with name mapping 1`] = `
"<div options={[{
  as: \\"foo\\",
  key: \\"bar\\"
}]} />;"
`;

exports[`transformStateOptions transforms a state getter with scope 1`] = `
"<div options={[{
  scope: \\"foo\\",
  key: \\"bar\\"
}]} />;"
`;

exports[`transformStateOptions transforms a state getter with scope and name mapping 1`] = `
"<div options={[{
  as: \\"aww\\",
  scope: \\"foo\\",
  key: \\"bar\\"
}]} />;"
`;

exports[`transformStateOptions transforms a state setter 1`] = `
"<div options={[{
  setter: \\"toggle\\",
  key: \\"foo\\"
}]} />;"
`;

exports[`transformStateOptions transforms a state setter with name mapping and scope 1`] = `
"<div options={[{
  as: \\"onClick\\",
  setter: \\"toggle\\",
  scope: \\"foo\\",
  key: \\"bar\\"
}]} />;"
`;

exports[`transformStateOptions transforms two state declarations 1`] = `
"<div options={[{
  as: \\"value\\",
  key: \\"name\\"
}, {
  as: \\"onChange\\",
  setter: \\"set\\",
  key: \\"name\\"
}]} />;"
`;
